### ISSUE TITLE
[Bug] Loner Influence

### ISSUE THREAD

**Issue by stergiosba** [stergiosba/PaxRL#2]
Currently when an agent is not a member of any neighbor, from now on called a loner, the steer acceleration is not zero. The following line shows why:

    return 20*(cohesion-X_dot) + 10*(alignment-X_dot) + 25*(separation-X_dot)

When cohesion=alignment=separation=0 then the steer should be zero but now it is -55*X_dot

**Comment by stergiosba** [comment]
Fixed by changing to this:

 enabled = jnp.any(neighbors_matrix,axis=0)[:,None]
 return enabled*(9*(cohesion-X_dot)+ 10*(alignment-X_dot) + 8*(separation-X_dot))

First line creates a vector that contains for every agent a boolean that checks if it has at least one neighbor.
[:.None] is needed so that matrix multiplication with a vector of shape (n,2) is possible.

**Comment by stergiosba** [comment]
Fixed by changing to this:

enabled = jnp.any(neighbors_matrix,axis=0)[:,None]
return enabled*(9*(cohesion-X_dot)+ 10*(alignment-X_dot) + 8*(separation-X_dot))

First line creates a vector that contains for every agent a boolean that checks if it has at least one neighbor.
[:.None] is needed so that matrix multiplication with a vector of shape (n,2) is possible.

**Comment by stergiosba** [comment]
The previous solution failed under the exreme case where the sole neighbor or an agent is the leader. In this case we still need the influence of the leader.

**Comment by stergiosba** [comment]
Possible solution is to multiply the result with the following mask:

    neighbors_mask = (jnp.any(neighbors_matrix,axis=0) + leader_neighbors_count)[:,None]
    return neighbors_mask *(20*(cohesion-X_dot) + 10*(alignment-X_dot) + 25*(separation-X_dot))

**Comment by stergiosba** [comment]
Fixed with this line:
```
return (total_mask*(w_c*cohesion+ w_a*alignment) + neighbors_mask*w_s*separation), leader
```

Issues for swarm model will be handled on a seperate repository. Closing this here.
